Mathlib/RingTheory/Kaehler/JacobiZariski.lean

```diff
@@ -0,0 +1,533 @@
+
+namespace Generators
+
+/--
+Given `R[X] → S` and `S[Y] → T`, this is the lift of an element in `ker(S[Y] → T)`
+to `ker(R[X][Y] → S[Y] → T)` constructed from `P.σ`.
+-/
+noncomputable
+def kerCompPreimage (x : Q.ker) :
+    (Q.comp P).ker := by
+  refine ⟨x.1.sum fun n r ↦ ?_, ?_⟩
+  · refine rename ?_ (P.σ r) * monomial ?_ 1
+    exacts [Sum.inr, n.mapDomain Sum.inl]
+  · simp only [ker_eq_ker_aeval_val, RingHom.mem_ker]
+    conv_rhs => rw [← show aeval Q.val x.1 = 0 from x.2, ← x.1.support_sum_monomial_coeff]
+    simp only [Finsupp.sum, map_sum, map_mul, aeval_rename, Function.comp_def, comp_val,
+      Sum.elim_inr, aeval_monomial, map_one, Finsupp.prod_mapDomain_index_inj Sum.inl_injective,
+      Sum.elim_inl, one_mul]
+    congr! with v i
+    simp_rw [← IsScalarTower.toAlgHom_apply R, ← comp_aeval, AlgHom.comp_apply, P.aeval_val_σ]
+    rfl
+
+lemma ofComp_kerCompPreimage (x : Q.ker) :
+    (Q.ofComp P).toAlgHom (kerCompPreimage Q P x) = x := by
+  conv_rhs => rw [← x.1.support_sum_monomial_coeff]
+  rw [kerCompPreimage, map_finsupp_sum, Finsupp.sum]
+  refine Finset.sum_congr rfl fun j _ ↦ ?_
+  simp only [AlgHom.toLinearMap_apply, _root_.map_mul, Hom.toAlgHom_monomial]
+  rw [one_smul, Finsupp.prod_mapDomain_index_inj Sum.inl_injective]
+  rw [rename, ← AlgHom.comp_apply, comp_aeval]
+  simp only [ofComp_val, Sum.elim_inr, Function.comp_apply, self_val, id_eq,
+    Sum.elim_inl, monomial_eq, Hom.toAlgHom_X]
+  congr 1
+  rw [aeval_def, IsScalarTower.algebraMap_eq R S, ← MvPolynomial.algebraMap_eq,
+    ← coe_eval₂Hom, ← map_aeval, P.aeval_val_σ]
+  rfl
+
+lemma Cotangent.map_ofComp_ker :
+    Submodule.map (Q.ofComp P).toAlgHom.toLinearMap ((Q.comp P).ker.restrictScalars R) =
+      Q.ker.restrictScalars R := by
+  apply le_antisymm
+  · rintro _ ⟨x, hx, rfl⟩
+    simp only [ker_eq_ker_aeval_val, Submodule.coe_restrictScalars, SetLike.mem_coe,
+      RingHom.mem_ker, AlgHom.toLinearMap_apply, Submodule.restrictScalars_mem] at hx ⊢
+    rw [← hx, Hom.algebraMap_toAlgHom]
+    rfl
+  · intro x hx
+    exact ⟨_, (kerCompPreimage Q P ⟨x, hx⟩).2, ofComp_kerCompPreimage Q P ⟨x, hx⟩⟩
+
+lemma Cotangent.surjective_map_ofComp :
+    Function.Surjective (Extension.Cotangent.map (Q.ofComp P).toExtensionHom) := by
+  intro x
+  obtain ⟨⟨x, hx⟩, rfl⟩ := Extension.Cotangent.mk_surjective x
+  have : x ∈ Q.ker.restrictScalars R := hx
+  rw [← map_ofComp_ker Q P] at this
+  obtain ⟨x, hx', rfl⟩ := this
+  exact ⟨.mk ⟨x, hx'⟩, Extension.Cotangent.map_mk _ _⟩
+
+lemma map_toComp_ker :
+    P.ker.map (Q.toComp P).toAlgHom.toRingHom = RingHom.ker (Q.ofComp P).toAlgHom := by
+  letI : DecidableEq (Q.vars →₀ ℕ) := Classical.decEq _
+  apply le_antisymm
+  · rw [Ideal.map_le_iff_le_comap]
+    rintro x (hx : algebraMap P.Ring S x = 0)
+    have : ((Q.ofComp P).toAlgHom.comp (Q.toComp P).toAlgHom) = IsScalarTower.toAlgHom R _ _ := by
+      ext1; simp
+    simp only [comp_vars, AlgHom.toRingHom_eq_coe, Ideal.mem_comap, RingHom.coe_coe,
```

Comment:
This proof looks quite intimidating (its ~70 LOC), but I also don't see a good way to cut it in pieces :/ Maybe you see one?

Reply:
I managed to extract one useful lemma but everything else seems quite specific and will probably not be used elsewhere.